packages/react/src/components/header/components/navigationLanguageSelector/NavigationLanguageSelector.tsx

```diff
@@ -2,67 +2,98 @@ import React, { PropsWithChildren, useEffect } from 'react';
     </button>
   );
 };
 
-const renderLanguageNode = (language: LanguageOption) => <LanguageButton key={language.value} language={language} />;
+const renderLanguageNode = (language: LanguageOption) => {
+  return <LanguageButton key={language.value} value={language.value} label={language.label} />;
+};
 
-export const NavigationLanguageSelector = ({ children, languages }: LanguageSelectorProps) => {
+export const NavigationLanguageSelector = ({ children, languages, fullWidthForMobile }: LanguageSelectorProps) => {
   const setAvailableLanguages = useSetAvailableLanguages();
-  const { isNotLargeScreen } = useHeaderContext();
+  const { isNotLargeScreen, languageSelectorContent } = useHeaderContext();
+  const { setLanguageSelectorContent } = useSetHeaderContext();
+
+  const isRender = !languages;
+  if (!isRender) {
+    useEffect(() => {
+      setAvailableLanguages(languages);
+    }, [languages]);
+  }
 
-  useEffect(() => {
-    setAvailableLanguages(languages);
-  }, [languages]);
+  if (children) {
+    useEffect(() => {
+      setLanguageSelectorContent(getChildElementsEvenIfContainersInbetween(children));
+    }, [children]);
+  }
 
-  const languageNodes = languages.map(renderLanguageNode);
+  const languageNodes = (languages || useAvailableLanguages()).map(renderLanguageNode);
+  // when its not large screen fullWidthForMobile -version can be used
+  const show =
+    isRender &&
+    ((isNotLargeScreen && fullWidthForMobile) || (!isNotLargeScreen && !fullWidthForMobile)) &&
+    Array.isArray(languageNodes) &&
+    languageNodes.length;
 
   return (
-    <div className={classes.languageSelector}>
-      {!isNotLargeScreen && languageNodes}
+    (show && (
+      <div className={classNames(classes.languageSelector, { [classes.fullWidthForMobile]: fullWidthForMobile })}>
+        <div className={classNames(classes.languageNodes)}>{languageNodes}</div>
 
-      <HeaderActionBarItemWithDropdown
-        id="language-selection-more"
-        iconClassName={classes.languageSelectorDropdownIcon}
-        dropdownClassName={classes.languageSelectorDropdown}
-        label={<IconAngleDown aria-hidden />}
-        closeLabel={<IconAngleUp aria-hidden />}
-        icon={IconGlobe}
-        closeIcon={IconGlobe}
-        fullWidth={isNotLargeScreen}
-        ariaLabel="select language"
-      >
-        {isNotLargeScreen && languageNodes}
-        {children}
-      </HeaderActionBarItemWithDropdown>
-    </div>
+        {languageSelectorContent && Array.isArray(languageSelectorContent) && languageSelectorContent.length ? (
+          <HeaderActionBarItemWithDropdown
+            id="language-selection-more"
+            iconClassName={classes.languageSelectorDropdownIcon}
+            dropdownClassName={classes.languageSelectorDropdown}
+            label={<IconAngleDown aria-hidden />}
+            closeLabel={<IconAngleUp aria-hidden />}
+            icon={<IconGlobe aria-hidden />}
+            closeIcon={<IconGlobe aria-hidden />}
```

Comment:
There's a bit too much spacing between the icons. Check the image here: left is this implementation and right is the [designs](https://www.figma.com/file/Kxwg3R0zNRHj55nQqFu6VS/HDS-UI-Kit-3.0?node-id=1012%3A22336&mode=dev).
<img width="195" alt="Screenshot 2023-08-16 at 8 14 48" src="https://github.com/City-of-Helsinki/helsinki-design-system/assets/108321134/07a7512c-4ac2-4377-b1c0-7a851da6763e">